File: src/spiral/managers/git.py
from __future__ import annotations

from spiral.managers.project import ProjectManager


class GitManager:
    def __init__(self, project: ProjectManager):
        self.project = project

    def _git(self, args: list[str]) -> str:
        result = self.project.run_command(["git", *args])
        return result.stdout + result.stderr

    def status(self) -> str:
        return self._git(["status", "--short"])

    def diff(self, staged: bool = True) -> str:
        args = ["diff"]
        if staged:
            args.append("--staged")
        return self._git(args)

    def add(self, paths: list[str] | None = None) -> str:
        args = ["add"]
        args.extend(paths if paths else ["."])
        return self._git(args)

    def commit(self, message: str) -> str:
        return self._git(["commit", "-m", message])

    def branch(self, action: str = "list", name: str | None = None) -> str:
        if action == "list":
            return self._git(["branch"])
        if action == "create" and name:
            return self._git(["branch", name])
        if action == "switch" and name:
            return self._git(["switch", name])
        return f"Error: unknown branch action '{action}'"

    def log(self, limit: int = 10) -> str:
        return self._git(["log", "--oneline", f"-{limit}"])

    def is_repo(self) -> bool:
        result = self.project.run_command(["git", "rev-parse", "--is-inside-work-tree"])
        return result.returncode == 0

File: src/managers/git.ts
import type { ProjectManager } from "./project.js";

export class GitManager {
  private project: ProjectManager;

  constructor(project: ProjectManager) {
    this.project = project;
  }

  private async git(args: string[]): Promise<string> {
    const { stdout, stderr } = await this.project.runCommand(["git", ...args]);
    return stdout + stderr;
  }

  async status(): Promise<string> {
    return this.git(["status", "--short"]);
  }

  async diff(staged = true): Promise<string> {
    const args = ["diff"];
    if (staged) args.push("--staged");
    return this.git(args);
  }

  async add(paths?: string[]): Promise<string> {
    const args = ["add", ...(paths ?? ["."])];
    return this.git(args);
  }

  async commit(message: string): Promise<string> {
    return this.git(["commit", "-m", message]);
  }

  async branch(action = "list", name?: string): Promise<string> {
    if (action === "list") return this.git(["branch"]);
    if (action === "create" && name) return this.git(["branch", name]);
    if (action === "switch" && name) return this.git(["switch", name]);
    return `Error: unknown branch action '${action}'`;
  }

  async log(limit = 10): Promise<string> {
    return this.git(["log", "--oneline", `-${limit}`]);
  }

  async isRepo(): Promise<boolean> {
    const { code } = await this.project.runCommand(["git", "rev-parse", "--is-inside-work-tree"]);
    return code === 0;
  }
}
